fv3core/decorators.py

```diff
@@ -96,502 +63,177 @@ def get_namespace(arg_specs, state):
-            self.device_sync = global_config.get_device_sync()
-        if self.device_sync is not None:
-            stencil_kwargs["device_sync"] = self.device_sync
 
     @property
-    def field_names(self) -> List[str]:
-        """Returns the list of stencil field names."""
-        return list(self.stencil_object.field_info.keys())
+    def field_names(self) -> Tuple[str]:
+        """names of stencil field call arguments"""
+        return cast(Tuple[str], tuple(self.stencil_object.field_info.keys()))
 
     @property
-    def parameter_names(self) -> List[str]:
-        """Returns the list of stencil parameter names."""
-        return list(self.stencil_object.parameter_info.keys())
-
-
-class FV3StencilObject(StencilWrapper):
-    """GT4Py stencil object used for fv3core."""
-
-    def __init__(self, func: Callable[..., None], **kwargs):
-        super().__init__(func, delay_compile=True)
-
-        self.times_called: int = 0
-        """Number of times this stencil has been called."""
-
-        self.timers = types.SimpleNamespace(call_run=0.0, run=0.0)
-        """Accumulated time spent in this stencil.
+    def parameter_names(self) -> Tuple[str]:
+        """names of stencil parameter call arguments"""
+        return cast(Tuple[str], tuple(self.stencil_object.parameter_info.keys()))
 
-        call_run includes stencil call overhead, while run omits it."""
 
-        self._passed_externals: Dict[str, Any] = kwargs.pop("externals", {})
-        """Externals passed in the decorator (others are added later)."""
-
-        self.backend_kwargs: Dict[str, Any] = kwargs
-        """Remainder of the arguments assumed to be compiler backend options."""
-
-        self._data_cache: StencilDataCache = StencilDataCache("data_cache.p")
-        """Data cache to store axis offsets and passed externals."""
+def compute_field_origins(
+    field_info_mapping, origin: Union[Index3D, Mapping[str, Tuple[int, ...]]]
+) -> Dict[str, Tuple[int, ...]]:
+    """Computes the origin for each field in the stencil call."""
+    if isinstance(origin, tuple):
+        field_origins: Dict[str, Tuple[int, ...]] = {"_all_": origin}
+        origin_tuple: Tuple[int, ...] = origin
+    else:
+        field_origins = {**origin}
+        origin_tuple = origin["_all_"]
+    field_names = tuple(field_info_mapping.keys())
+    for i, field_name in enumerate(field_names):
+        if field_name not in field_origins:
+            field_info = field_info_mapping[field_name]
+            if field_info is not None:
+                field_origin_list = []
+                for ax in field_info.axes:
+                    origin_index = {"I": 0, "J": 1, "K": 2}[ax]
+                    field_origin_list.append(origin_tuple[origin_index])
+                field_origin = tuple(field_origin_list)
+            else:
+                field_origin = origin_tuple
+            field_origins[field_name] = field_origin
+    return field_origins
```

Comment:
If `stencil_object._make_origin_dict` is called before this call (or `stencil_object` is passed in and called), then the following code can replace this and will be very close to what gt4py does. 

Note that code like `origin_index = {"I": 0, "J": 1, "K": 2}[ax]` is not guaranteed to aways work.

```python
    field_origins = stencil_object._make_origin_dict(origin)  # optional: if not done before call
    all_origin = field_origins.get("_all_", None)

    for field_name, field_info in field_info_mapping.items():
        if field_info is not None:
            field_origin = field_origins.get(name, None)
            if field_origin is not None:
                field_origin_ndim = len(field_origin)
                if field_origin_ndim != field_info.ndim:
                    assert field_origin_ndim == field_info.domain_ndim, (
                        f"Invalid origin specification ({field_origin})"
                        f" for '{name}' field."
                    )
                    field_origins[name] = (
                        *field_origin,
                        *((0,) * len(field_info.data_dims)),
                    )
            elif all_origin is not None:
                field_origins[name] = (
                    *gt_utils.filter_mask(all_origin, field_info.domain_mask),
                    *((0,) * len(field_info.data_dims)),
                )
            else:
                raise ValueError(f"Missing origin for {name}")
    return field_origins
```

Reply:
Actually, now that I think about it I believe this requires the latest gt4py, so hold off for now and I'll change it after.